### PR TITLE
Building a Job Board Platform with Directus and SolidStart.js

### PR DESCRIPTION
**Pull request by preshenv**
This is the first draft for issue https://github.com/directus-labs/guest-authoring/issues/186

### CODE REVIEW

File: directus-job-board-platform/index.md [hunk 1]
```diff
@@ -0,0 +1,1126 @@
+---
+title: 'Building a Job Board Platform with Directus and SolidStart.js'
+description: 'Learn how to build a job board portal with Directus and SolidStart.js. You'll learn how to register new users, login, and perform create, read, update, and delete operations (CRUD) on Directus.'
+author:
+  name: 'Ndoma Precious'
+  avatar_file_name: './ndoma-precious.png'
+---
+
+In this tutorial, you'll learn to build a job board portal using Directus and SolidStart.js. We'll cover user registration, login, and CRUD operations with Directus. You'll create a complete job board with listing, application, and management features for both jobs and applications. This guide will provide you with the skills to combine Directus's backend capabilities with SolidStart.js's reactive frontend, to build a job portal web application.
```
Comment (phazonoverload) [inline directus-job-board-platform/index.md:9]:
```suggestion
In this tutorial, you'll learn to build a job board portal using Directus and SolidStart.js. We'll cover user registration, login, and working with data in Directus. You'll create a complete job board with listing, application, and management features for both jobs and applications. This guide will provide you with the skills to combine Directus's backend capabilities with SolidStart.js's reactive frontend, to build a job portal web application.
```

File: directus-job-board-platform/index.md [hunk 2]
```diff
@@ -0,0 +1,1126 @@
...
+
+## Before You Start
+You will need:
+
+- [Node.js v18](https://nodejs.org/) and above installed on your computer.
```
Comment (phazonoverload) [inline directus-job-board-platform/index.md:14]:
```suggestion
- [Node.js v18](https://nodejs.org/) or above installed on your computer.
```

File: directus-job-board-platform/index.md [hunk 3]
```diff
@@ -0,0 +1,1126 @@
...
+3. Choose `first_name` as the Display template.
+
+This creates a relationship between the job and the admin who created it.
+
+In the `application` collection, add two Many-to-One fields:
+
+1. `user`: Links to the applicant
+2. `job`: Links to the job
+
+These fields connect each application to its corresponding applicant and job. Add 3 items in the `job` collection - [here's some sample data](https://raw.githubusercontent.com/preshenv/directus-solidstart-job-board/main/src/components/jobs.md).
```
Comment (phazonoverload) [inline directus-job-board-platform/index.md:49]:
```suggestion
## Configuring Directus Data Models

Create a `job`, and `application` collection in your project. The `job` will store a list available jobs with the following fields:

- `id`: autocomplete input
- `title`: input field
- `location`: input field
- `type`: input field
- `salary`: input field

The `application` collection will store job applications with the following fields:

- `id:` Autocomplete Input
- `status`: input field

To link a new job with its creator:

1. In the `job` collection, add a Many-to-One field named `employer`.
2. Set the related collection to System > `directus_users`.
3. Choose `first_name` as the Display template.

This creates a relationship between the job and the admin who created it.

In the `application` collection, add two Many-to-One fields:

1. `user`: Links to the applicant
2. `job`: Links to the job

These fields connect each application to its corresponding applicant and job. Add 3 items in the `job` collection - [here's some sample data](https://raw.githubusercontent.com/preshenv/directus-solidstart-job-board/main/src/components/jobs.md).
```

File: directus-job-board-platform/index.md [hunk 4]
```diff
@@ -0,0 +1,1126 @@
...
+In your access control settings, create a new role called `Job Applicant`. For the `application` collection, enable `create` and `read` permissions. Use custom rules for the `application` collection to ensure users can only read and update their own applications. Set a filter like: `user -> id Equals $CURRENT_USER.id`. This allows users to view all jobs, create new applications, and view or update only their own applications. 
+
+## Allowing Public Registration
+Open the public role, find `directus_users` under system collections, and then open custom permissions for the `create` operation. 
+1. In field permissions, enable only `first_name`, `last_name`, `email`, and `password` options. If you want users to provide other data at the time of registration, enable the respective field.
+2. In field validation, require `first_name`, `last_name`, `email`, and `password` to not be empty.
+3. In field presets, set the new user's `role` to the `Job Applicant` role id. 
+
+Give public view access to the `job` collection to allow users to see available jobs even when they are not logged in. Also, ensure the admin role retains full permissions across all collections.
```
Comment (phazonoverload) [inline directus-job-board-platform/index.md:60]: Please rework this section to use our new user registration endpoint. It must first be enabled in project settings, and then an unauthenticated request can be made to POST /users/register: https://docs.directus.io/reference/authentication.html#register

File: directus-job-board-platform/index.md [hunk 5]
```diff
@@ -0,0 +1,1126 @@
...
+    .with(rest({ credentials: "include" }));
+  return directus;
+}
+export default getDirectusInstance;
+```
+Add your Directus URL to the `.env` file:
+
+```env
+VITE_PUBLIC_DIRECTUS_API_URL='https://directus.example.com'
+```
```
Comment (phazonoverload) [inline directus-job-board-platform/index.md:93]:
````suggestion
## Initializing a SolidStart.js project

Create a new SolidStart project by running the command:

```env
npm init solid@latest
```

Choose the **bare** template, enable Server-side rendering from the prompt, and Typescript in the prompts.

In your SolidStart project's `src` directory, create a `lib` directory. Inside it, create a `directus.js` file:

```jsx
import { authentication, createDirectus, rest } from "@directus/sdk";

export const PUBLIC_DIRECTUS_API_URL = import.meta.env
  .VITE_PUBLIC_DIRECTUS_API_URL;

function getDirectusInstance() {
  const directus = createDirectus(PUBLIC_DIRECTUS_API_URL)
    .with(
      authentication("cookie", { credentials: "include", autoRefresh: true })
    )
    .with(rest({ credentials: "include" }));
  return directus;
}
export default getDirectusInstance;
```

Add your Directus URL to the `.env` file:

```env
VITE_PUBLIC_DIRECTUS_API_URL='https://directus.example.com'
```
````

Consolidating as it's all part of setup

File: directus-job-board-platform/index.md [hunk 6]
```diff
@@ -0,0 +1,1126 @@
...
+
+## Implementing User Authentication
+To implement user authentication to grant users access to the application, create a **context**, and inside that folder create an `AuthContext.tsx` file and add the following code:
```
Comment (phazonoverload) [inline directus-job-board-platform/index.md:96]: > create a **context**

A context... directory?

Comment (phazonoverload) [reply directus-job-board-platform/index.md:96]: This whole section needs subheadings focusing on each part of the authentication - register, login, fetching the user, logout, etc. This is too much all at once - break it down for comprehension.

File: directus-job-board-platform/index.md [hunk 7]
```diff
@@ -0,0 +1,1126 @@
...
+      <Router root={props => <Suspense>{props.children}</Suspense>}>
+        <FileRoutes />
+      </Router>
+    </AuthProvider>
+  );
+}
+```
+This setup enables automatic routing based on your file structure while providing authentication context throughout the app.
+
+## Creating components for displaying job listings
```
Comment (phazonoverload) [inline directus-job-board-platform/index.md:402]:
```suggestion
## Creating Job Listings Components
```

File: directus-job-board-platform/index.md [hunk 8]
```diff
@@ -0,0 +1,1126 @@
...
+      </Show>
+    </div>
+  );
+}
+
+export default HomePage;
+```
+![Job Listing Portal](<Screenshot 2024-07-02 at 10.43.05.png>)
+
+## Creating, updating, and deleting job listings
```
Comment (phazonoverload) [inline directus-job-board-platform/index.md:518]:
```suggestion
## Creating, Updating, and Deleting Listings
```

Please also break this down into i's constituent parts. It's too much to do at the same time.

File: directus-job-board-platform/index.md [hunk 9]
```diff
@@ -0,0 +1,1126 @@
...
+  );
+}
+```
+Create a new file named `Modal.css` in your **components** and copy the CSS styles [here](https://github.com/preshenv/directus-solidstart-job-board/blob/main/src/components/Modal.css)to it.
+
+Log in with admin credentials, and click on the **Add New Job** button to create a new job, you can also edit and delete a job by clicking on the edit and delete buttons respectively
+
+![Add new job modal](<Screenshot 2024-07-02 at 12.01.27.png>)
+
+## Implementing search and filtering functionality
```
Comment (phazonoverload) [inline directus-job-board-platform/index.md:763]:
```suggestion
## Implementing Search and Filters
```

**Review by phazonoverload — CHANGES_REQUESTED** [review]
Good start, but a couple of recurring notes:

1. Both the auth and the create/update/delete sections need breaking up. Once you preset a code snippet past a certain size, it's hard for readers to gain understanding from it. 
2. Whenever mentioning a directory, please wrap the directory name in backticks instead of double quotes. 
3. Please remove all styling and CSS from this tutorial as it's distracting from the key learnings. This means removing classes from elements where they are only used for styling, and possibly also the whole modal, opting instead for a short page.

File: directus-job-board-platform/index.md (hunk 6 above)
Comment (preshenv) [reply directus-job-board-platform/index.md]: resolved

File: directus-job-board-platform/index.md (hunk 4 above)
Comment (preshenv) [reply directus-job-board-platform/index.md]: resolved

**Comment by preshenv** [comment]
Hi @phazonoverload 
I have resolved and updated the draft to add ur commit messages.

**Comment by phazonoverload** [comment]
Thank you for writing this, and we are happy to consider it done in it's current state. My team will schedule it in to be published. 

This repo is just for authoring/reviewing - the actual piece will be published via a Directus project, so we'll take it from here. I'll close this issue to keep the main branch clean, but this is a acceptance of your work.

Could you email us at [devrel@directus.io](mailto:devrel@directus.io) and we'll share what we need to get you paid. Thanks!

**Pull request closed without merging**